### ISSUE TITLE
Error denormalizing to non-cloneable objects

### ISSUE THREAD

**Issue by colinodell** [api-platform/core#5561]
**API Platform version(s) affected**: 2.7.0 through 3.1.8

**Description**  
`ApiPlatform\Serializer\AbstractItemNormalizer::denormalize()` attempts to clone objects without checking if they're actually cloneable, resulting in a fatal error.

https://github.com/api-platform/core/blob/3de661c2d163629cb9097de5a0370612a00f6523/src/Serializer/AbstractItemNormalizer.php#L217

**How to reproduce**  

Attempt to denormalize any object where `object_to_populate` has a `__clone()` function that is `private` or `protected`, like this:

```php
class CannotBeCloned {
    public string $foo;

    private function __clone()
    {
    }
};

$original = new CannotBeCloned();

$itemNormalizer->denormalize(['foo' => 'bar'], CannotBeCloned::class, context: ['resource_class' => CannotBeCloned::class, 'object_to_populate' => $original]);
```

Result:

    Fatal error: Uncaught Error: Call to private CannotBeCloned::__clone() from scope ApiPlatform\Serializer\AbstractItemNormalizer

**Possible Solution**  

Utilize [your `CloneTrait`](https://github.com/api-platform/core/blob/v3.1.8/src/Util/CloneTrait.php) to avoid this fatal error, and then perhaps skip the post-denormalize-revert loop for non-cloneable objects?  Or whatever you feel is most appropriate.

**Additional Context**  

It seems this BC break was introduced in v2.7.0 to support the `security_post_denormalize` feature (#4184).  We discovered this trying to upgrade from 2.6.8 to any newer version.

**Comment by soyuka** [comment]
Interesting, and thanks for the nice bug report. I don't really want to maintain 2.7 but if you provide a fix I'll tag a new release. One thing is sure we need a fix on 3.1 anyway.

**Comment by colinodell** [comment]
I totally understand :) Personally, I'll need to write a patch for 2.7 regardless (for my own usage), so I'd be happy to contribute that as a PR too.
